lib/portlayer/vspc/vspc.go

```diff
@@ -0,0 +1,230 @@
+	KnownSuboptions2      byte          = 1
+	UnknownSuboptionRcvd1 byte          = 2
+	UnknownSuboptionRcvd2 byte          = 3
+	VmotionBegin          byte          = 40
+	VmotionGoahead        byte          = 41
+	VmotionNotNow         byte          = 43
+	VmotionPeer           byte          = 44
+	VmotionPeerOK         byte          = 45
+	VmotionComplete       byte          = 46
+	VmotionAbort          byte          = 48
+	DoProxy               byte          = 70
+	WillProxy             byte          = 71
+	WontProxy             byte          = 73
+	VMVCUUID              byte          = 80
+	GetVMVCUUID           byte          = 81
+	VMName                byte          = 82
+	GetVMName             byte          = 83
+	VMBiosUUID            byte          = 84
+	GetVMBiosUUID         byte          = 85
+	VMLocationUUID        byte          = 86
+	GetsVMLocationUUID    byte          = 87
+	VMHealthCheckPeriod   time.Duration = 10 * time.Second
+	vspcPort              int           = 2377
+)
+const remoteConnReadDeadline = 1 * time.Second
+
+// VM is a struct that represents a VM
+type VM struct {
+	sync.Mutex
+	// this is the current connection between the vspc and the containerVM serial port
+	containerConn *telnet.Conn
+	// in case of vmotion, at some point we will have two telnet connections to the containerVM
+	// prevContainerConn is the connection of the source
+	// containerConn will be the connection to the destination
+	prevContainerConn *telnet.Conn
+	// inVmotion is a boolean denoting whether this VM is in a state of vmotion or not
+	inVmotion bool
+
+	// remoteConn is the remote-system connection
+	// this is the connection between the vspc and the attach-server
+	remoteConn net.Conn
+
+	vmUUID           string
+	vmotionStarted   chan chan struct{}
+	vmotionCompleted chan chan struct{}
+}
+
+// NewVM is the constructor of the VM
+func NewVM(tc *telnet.Conn) *VM {
+	return &VM{
+		containerConn:    tc,
+		inVmotion:        false,
+		vmotionStarted:   make(chan chan struct{}),
+		vmotionCompleted: make(chan chan struct{}),
+	}
+}
+
+// Vspc is all the vspc singletons
+type Vspc struct {
+	vmManagerMu *sync.Mutex
+	vmManager   map[string]*VM
+
+	*telnet.Server
+
+	attachSrvrAddr string
+	attachSrvrPort uint
+
+	doneCh chan bool
+}
+
+// NewVspc is the constructor
+func NewVspc(address string, port uint, attachSrvrAddr string, attachSrvrPort uint, ch chan bool) *Vspc {
+	vspc := &Vspc{
+		vmManagerMu: new(sync.Mutex),
+		vmManager:   make(map[string]*VM),
+
+		attachSrvrAddr: attachSrvrAddr,
+		attachSrvrPort: attachSrvrPort,
+
+		doneCh: ch,
+	}
+	hdlr := Handler{vspc}
+	opts := telnet.ServerOpts{
+		Addr:        fmt.Sprintf("%s:%d", address, port),
+		ServerOpts:  []byte{telnet.Binary, telnet.Sga, telnet.Echo},
+		ClientOpts:  []byte{telnet.Binary, telnet.Sga, VmwareExt},
+		DataHandler: hdlr.DataHdlr,
+		CmdHandler:  hdlr.CmdHdlr,
+	}
+	vspc.Server = telnet.NewServer(opts)
+	go vspc.monitorVMConnections()
```

Comment:
IMHO NewX shouldn't take any action but just return the said struct. This could be a Start/Monitor etc. method on Vspc